File: client_utils/src/math.rs
//! Optional basic math types and utilities for `plaza_client_utils`.
//!
//! These are provided for convenience if an application does not want to bring in
//! a larger math library, or for default implementations of traits like
//! `Interpolatable` and `Extrapolatable` on these basic types.
//!
//! Applications are encouraged to use their own preferred math libraries and implement
//! the necessary traits (`Interpolatable`, `Extrapolatable`) for their chosen types.

use crate::extrapolation::Extrapolatable;
use crate::interpolation::Interpolatable;
use std::fmt::Debug;
use std::ops::{Add, Div, Mul, Sub};


#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Vec2 {
  pub x: f32,
  pub y: f32,
}

impl Vec2 {
  pub const ZERO: Self = Self { x: 0.0, y: 0.0 };
  pub const ONE: Self = Self { x: 1.0, y: 1.0 };

  pub fn new(x: f32, y: f32) -> Self {
    Self { x, y }
  }

  pub fn length_squared(self) -> f32 {
    self.x * self.x + self.y * self.y
  }

  pub fn length(self) -> f32 {
    self.length_squared().sqrt()
  }

  pub fn normalize(self) -> Self {
    let len = self.length();
    if len > f32::EPSILON {
      self / len
    } else {
      Self::ZERO // Avoid division by zero
    }
  }
}

impl Add for Vec2 {
  type Output = Self;
  fn add(self, rhs: Self) -> Self::Output {
    Self {
      x: self.x + rhs.x,
      y: self.y + rhs.y,
    }
  }
}

impl Sub for Vec2 {
  type Output = Self;
  fn sub(self, rhs: Self) -> Self::Output {
    Self {
      x: self.x - rhs.x,
      y: self.y - rhs.y,
    }
  }
}

impl Mul<f32> for Vec2 {
  type Output = Self;
  fn mul(self, rhs: f32) -> Self::Output {
    Self {
      x: self.x * rhs,
      y: self.y * rhs,
    }
  }
}

impl Div<f32> for Vec2 {
  type Output = Self;
  fn div(self, rhs: f32) -> Self::Output {
    Self {
      x: self.x / rhs,
      y: self.y / rhs,
    } // Panics if rhs is 0
  }
}


#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Vec3 {
  pub x: f32,
  pub y: f32,
  pub z: f32,
}

impl Vec3 {
  pub const ZERO: Self = Self { x: 0.0, y: 0.0, z: 0.0 };
  pub const ONE: Self = Self { x: 1.0, y: 1.0, z: 1.0 };

  pub fn new(x: f32, y: f32, z: f32) -> Self {
    Self { x, y, z }
  }
  // ... (length_squared, length, normalize - similar to Vec2) ...
  pub fn length_squared(self) -> f32 {
    self.x * self.x + self.y * self.y + self.z * self.z
  }

  pub fn length(self) -> f32 {
    self.length_squared().sqrt()
  }

  pub fn normalize(self) -> Self {
    let len = self.length();
    if len > f32::EPSILON {
      self / len
    } else {
      Self::ZERO
    }
  }
}

impl Add for Vec3 {
  type Output = Self;
  fn add(self, rhs: Self) -> Self::Output {
    Self {
      x: self.x + rhs.x,
      y: self.y + rhs.y,
      z: self.z + rhs.z,
    }
  }
}

impl Sub for Vec3 {
  type Output = Self;
  fn sub(self, rhs: Self) -> Self::Output {
    Self {
      x: self.x - rhs.x,
      y: self.y - rhs.y,
      z: self.z - rhs.z,
    }
  }
}

impl Mul<f32> for Vec3 {
  type Output = Self;
  fn mul(self, rhs: f32) -> Self::Output {
    Self {
      x: self.x * rhs,
      y: self.y * rhs,
      z: self.z * rhs,
    }
  }
}

impl Div<f32> for Vec3 {
  type Output = Self;
  fn div(self, rhs: f32) -> Self::Output {
    Self {
      x: self.x / rhs,
      y: self.y / rhs,
      z: self.z / rhs,
    }
  }
}

// Basic implementation for slerp example. A real quaternion would have more methods.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Quat {
  pub x: f32,
  pub y: f32,
  pub z: f32,
  pub w: f32,
}

impl Quat {
  pub const IDENTITY: Self = Self {
    x: 0.0,
    y: 0.0,
    z: 0.0,
    w: 1.0,
  };

  pub fn new(x: f32, y: f32, z: f32, w: f32) -> Self {
    Self { x, y, z, w }
  }

  pub fn dot(self, other: Self) -> f32 {
    self.x * other.x + self.y * other.y + self.z * other.z + self.w * other.w
  }

  pub fn normalize(self) -> Self {
    let mag_sq = self.dot(self);
    if mag_sq > f32::EPSILON {
      let mag = mag_sq.sqrt();
      Self {
        x: self.x / mag,
        y: self.y / mag,
        z: self.z / mag,
        w: self.w / mag,
      }
    } else {
      Self::IDENTITY // Or panic, undefined normalization
    }
  }

  // Basic slerp implementation
  pub fn slerp(self, mut end: Self, t: f32) -> Self {
    let mut dot = self.dot(end);

    // If the dot product is negative, the quaternions are more than 90 degrees apart,
    // so we should invert one to take the shorter path.
    if dot < 0.0 {
      end = Quat::new(-end.x, -end.y, -end.z, -end.w);
      dot = -dot;
    }

    const DOT_THRESHOLD: f32 = 0.9995; // Threshold for linear interpolation
    if dot > DOT_THRESHOLD {
      // Quaternions are very close, use linear interpolation (and then normalize)
      // to avoid issues with sin(angle).
      let result = Quat::new(
        self.x + t * (end.x - self.x),
        self.y + t * (end.y - self.y),
        self.z + t * (end.z - self.z),
        self.w + t * (end.w - self.w),
      );
      result.normalize()
    } else {
      // Standard slerp
      let theta_0 = dot.acos(); // angle between input vectors
      let theta = theta_0 * t; // angle between v0 and result
      let sin_theta = theta.sin();
      let sin_theta_0 = theta_0.sin();

      let s0 = (theta_0 - theta).cos() - dot * sin_theta / sin_theta_0; // == theta.cos()
      let s1 = sin_theta / sin_theta_0;

      Quat::new(
        (s0 * self.x) + (s1 * end.x),
        (s0 * self.y) + (s1 * end.y),
        (s0 * self.z) + (s1 * end.z),
        (s0 * self.w) + (s1 * end.w),
      )
      .normalize() // Ensure result is normalized due to potential float inaccuracies
    }
  }

  /// Hamilton product: composes two rotations.
  pub fn multiply(self, rhs: Self) -> Self {
    Quat {
      w: self.w * rhs.w - self.x * rhs.x - self.y * rhs.y - self.z * rhs.z,
      x: self.w * rhs.x + self.x * rhs.w + self.y * rhs.z - self.z * rhs.y,
      y: self.w * rhs.y - self.x * rhs.z + self.y * rhs.w + self.z * rhs.x,
      z: self.w * rhs.z + self.x * rhs.y - self.y * rhs.x + self.z * rhs.w,
    }
  }
}


impl<Timestamp> Interpolatable<Timestamp> for f32
where
  Timestamp: Copy + Debug + PartialOrd + Sub<Output = Timestamp> + crate::interpolation::ToF32,
{
  fn interpolate(&self, other: &Self, t: f32, _time_a: Timestamp, _time_b: Timestamp) -> Self {
    self + (other - self) * t
  }
}

impl<Timestamp> Interpolatable<Timestamp> for Vec2
where
  Timestamp: Copy + Debug + PartialOrd + Sub<Output = Timestamp> + crate::interpolation::ToF32,
{
  fn interpolate(&self, other: &Self, t: f32, _time_a: Timestamp, _time_b: Timestamp) -> Self {
    Vec2 {
      x: self.x.interpolate(&other.x, t, _time_a, _time_b),
      y: self.y.interpolate(&other.y, t, _time_a, _time_b),
    }
  }
}

impl<Timestamp> Interpolatable<Timestamp> for Vec3
where
  Timestamp: Copy + Debug + PartialOrd + Sub<Output = Timestamp> + crate::interpolation::ToF32,
{
  fn interpolate(&self, other: &Self, t: f32, _time_a: Timestamp, _time_b: Timestamp) -> Self {
    Vec3 {
      x: self.x.interpolate(&other.x, t, _time_a, _time_b),
      y: self.y.interpolate(&other.y, t, _time_a, _time_b),
      z: self.z.interpolate(&other.z, t, _time_a, _time_b),
    }
  }
}

impl<Timestamp> Interpolatable<Timestamp> for Quat
where
  Timestamp: Copy + Debug + PartialOrd + Sub<Output = Timestamp> + crate::interpolation::ToF32,
{
  fn interpolate(&self, other: &Self, t: f32, _time_a: Timestamp, _time_b: Timestamp) -> Self {
    self.slerp(*other, t)
  }
}

// Velocity types will be Vec2 for Vec2 position, Vec3 for Vec3 position.

// Example: Extrapolating Vec3 position with Vec3 linear velocity
// TimeDelta is assumed to be f32 seconds here for simplicity.
impl Extrapolatable<Vec3, f32> for Vec3 {
  fn extrapolate_with_velocity(&self, velocity: &Vec3, delta_time_secs: f32) -> Self {
    *self + (*velocity * delta_time_secs)
  }
}

// Example: Extrapolating Quat rotation.
// This is more complex. A common way is to treat angular_velocity as an axis-angle vector
// (axis is direction, magnitude is radians per second). Convert this to a delta quaternion
// for the delta_time_secs, then multiply the current rotation by this delta.
// Or that angular_velocity is represented as another Quat (a small rotation).

// This is a very simplified angular extrapolation, assuming Velocity is a delta Quat for that step.
// Deliberately minimal: the "velocity" is taken as a per-second delta rotation
// and slerped toward by `delta_time_secs`. Real angular velocity is a Vec3
// axis-angle, which this minimal Quat cannot express; for real rotational
// extrapolation, implement `Extrapolatable` for your engine's quaternion type.
impl Extrapolatable<Quat, f32> for Quat {
  fn extrapolate_with_velocity(&self, angular_velocity_as_delta_quat_per_sec: &Quat, delta_time_secs: f32) -> Self {
    if angular_velocity_as_delta_quat_per_sec.w.abs() < 1.0 - f32::EPSILON {
      let scaled_delta_rotation = Quat::IDENTITY.slerp(*angular_velocity_as_delta_quat_per_sec, delta_time_secs);
      return self.multiply(scaled_delta_rotation).normalize();
    }
    // An identity-like velocity rotates nothing.
    *self
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  
  use crate::types::ClientTimeMs;

  #[test]
  fn vec2_ops() {
    let v1 = Vec2::new(1.0, 2.0);
    let v2 = Vec2::new(3.0, 4.0);
    assert_eq!(v1 + v2, Vec2::new(4.0, 6.0));
    assert_eq!(v2 - v1, Vec2::new(2.0, 2.0));
    assert_eq!(v1 * 2.0, Vec2::new(2.0, 4.0));
    assert_eq!(v2 / 2.0, Vec2::new(1.5, 2.0));
  }

  #[test]
  fn vec3_ops() {
    let v1 = Vec3::new(1.0, 2.0, 3.0);
    let v2 = Vec3::new(4.0, 5.0, 6.0);
    assert_eq!(v1 + v2, Vec3::new(5.0, 7.0, 9.0));
    // ... more tests ...
  }

  #[test]
  fn f32_interpolate() {
    let a = 10.0f32;
    let b = 20.0f32;
    let time_a: ClientTimeMs = 0;
    let time_b: ClientTimeMs = 100;
    assert!((a.interpolate(&b, 0.0, time_a, time_b) - 10.0).abs() < f32::EPSILON);
    assert!((a.interpolate(&b, 0.5, time_a, time_b) - 15.0).abs() < f32::EPSILON);
    assert!((a.interpolate(&b, 1.0, time_a, time_b) - 20.0).abs() < f32::EPSILON);
  }

  #[test]
  fn vec3_interpolate() {
    let v_a = Vec3::new(0.0, 10.0, 20.0);
    let v_b = Vec3::new(10.0, 20.0, 20.0);
    let time_a: ClientTimeMs = 0;
    let time_b: ClientTimeMs = 100;

    let v_mid = v_a.interpolate(&v_b, 0.5, time_a, time_b);
    assert_eq!(v_mid, Vec3::new(5.0, 15.0, 20.0));
  }

  #[test]
  fn quat_slerp_identity() {
    let q_ident = Quat::IDENTITY;
    let q_other = Quat::new(0.0, 1.0, 0.0, 0.0).normalize(); // 180 deg rot around Y

    let q_mid = q_ident.slerp(q_other, 0.5);
    // Should be 90 deg rot around Y, approx (0, 0.707, 0, 0.707)
    assert!((q_mid.y - std::f32::consts::FRAC_1_SQRT_2).abs() < 1e-5);
    assert!((q_mid.w - std::f32::consts::FRAC_1_SQRT_2).abs() < 1e-5);
    assert!(q_mid.x.abs() < 1e-5);
    assert!(q_mid.z.abs() < 1e-5);
  }

  #[test]
  fn vec3_extrapolate() {
    let pos = Vec3::new(1.0, 0.0, 0.0);
    let vel = Vec3::new(2.0, 0.0, 0.0); // Moving 2 units/sec on X
    let dt = 0.5_f32; // Extrapolate for 0.5 seconds

    let next_pos = pos.extrapolate_with_velocity(&vel, dt);
    assert_eq!(next_pos, Vec3::new(2.0, 0.0, 0.0)); // 1.0 + 2.0 * 0.5 = 2.0
  }
}

File: flutter/plaza_client_utils/lib/src/math.dart
/// Optional basic math types for `plaza_client_utils`.
///
/// Provided for convenience if an application does not want to bring in a larger
/// math library, and as the interpolate and extrapolate rules for these basic
/// types. Applications are encouraged to use their own preferred math library:
/// pass its lerp and extrapolate functions to the primitives that take them.
///
/// A Flutter or Flame application already has `vector_math`, whose `Vector2` is
/// mutable and better integrated with everything around it. These exist so this
/// package can stay dependency-free, not to compete with it.
///
/// Ported from `plaza_client_utils::math`.
library;

import 'dart:math' as math;

/// The smallest positive value that changes a `double` near 1.0.
///
/// The Rust original guards its normalisations with `f32::EPSILON`. Dart has no
/// float32, so this is the `double` equivalent, which makes the guard tighter
/// rather than looser.
const double doubleEpsilon = 2.220446049250313e-16;

class Vec2 {
  const Vec2(this.x, this.y);

  static const Vec2 zero = Vec2(0.0, 0.0);
  static const Vec2 one = Vec2(1.0, 1.0);

  final double x;
  final double y;

  double get lengthSquared => x * x + y * y;
  double get length => math.sqrt(lengthSquared);

  /// The unit vector in this direction, or [zero] for a vector too short to have
  /// a direction.
  Vec2 normalize() {
    final len = length;
    return len > doubleEpsilon ? this / len : Vec2.zero;
  }

  Vec2 operator +(Vec2 rhs) => Vec2(x + rhs.x, y + rhs.y);
  Vec2 operator -(Vec2 rhs) => Vec2(x - rhs.x, y - rhs.y);
  Vec2 operator *(double rhs) => Vec2(x * rhs, y * rhs);
  Vec2 operator /(double rhs) => Vec2(x / rhs, y / rhs);
  Vec2 operator -() => Vec2(-x, -y);

  /// Straight-line blend toward [other], the rule to hand a `SnapshotBuffer`.
  Vec2 lerp(Vec2 other, double t) => Vec2(x + (other.x - x) * t, y + (other.y - y) * t);

  /// Coasts along [velocity] for [dtSecs], the rule to hand a `RemoteView`.
  Vec2 extrapolate(Vec2 velocity, double dtSecs) => this + velocity * dtSecs;

  @override
  bool operator ==(Object other) => other is Vec2 && other.x == x && other.y == y;

  @override
  int get hashCode => Object.hash(x, y);

  @override
  String toString() => 'Vec2($x, $y)';
}

class Vec3 {
  const Vec3(this.x, this.y, this.z);

  static const Vec3 zero = Vec3(0.0, 0.0, 0.0);
  static const Vec3 one = Vec3(1.0, 1.0, 1.0);

  final double x;
  final double y;
  final double z;

  double get lengthSquared => x * x + y * y + z * z;
  double get length => math.sqrt(lengthSquared);

  Vec3 normalize() {
    final len = length;
    return len > doubleEpsilon ? this / len : Vec3.zero;
  }

  Vec3 operator +(Vec3 rhs) => Vec3(x + rhs.x, y + rhs.y, z + rhs.z);
  Vec3 operator -(Vec3 rhs) => Vec3(x - rhs.x, y - rhs.y, z - rhs.z);
  Vec3 operator *(double rhs) => Vec3(x * rhs, y * rhs, z * rhs);
  Vec3 operator /(double rhs) => Vec3(x / rhs, y / rhs, z / rhs);
  Vec3 operator -() => Vec3(-x, -y, -z);

  Vec3 lerp(Vec3 other, double t) =>
      Vec3(x + (other.x - x) * t, y + (other.y - y) * t, z + (other.z - z) * t);

  Vec3 extrapolate(Vec3 velocity, double dtSecs) => this + velocity * dtSecs;

  @override
  bool operator ==(Object other) => other is Vec3 && other.x == x && other.y == y && other.z == z;

  @override
  int get hashCode => Object.hash(x, y, z);

  @override
  String toString() => 'Vec3($x, $y, $z)';
}

/// A minimal quaternion, enough to slerp a rotation.
class Quat {
  const Quat(this.x, this.y, this.z, this.w);

  static const Quat identity = Quat(0.0, 0.0, 0.0, 1.0);

  final double x;
  final double y;
  final double z;
  final double w;

  double dot(Quat other) => x * other.x + y * other.y + z * other.z + w * other.w;

  Quat normalize() {
    final magSq = dot(this);
    if (magSq <= doubleEpsilon) return Quat.identity;
    final mag = math.sqrt(magSq);
    return Quat(x / mag, y / mag, z / mag, w / mag);
  }

  /// Spherical interpolation toward [end], taking the shorter arc.
  Quat slerp(Quat end, double t) {
    var target = end;
    var d = dot(target);

    // A negative dot means they are more than 90 degrees apart, so invert one to
    // take the shorter path.
    if (d < 0.0) {
      target = Quat(-target.x, -target.y, -target.z, -target.w);
      d = -d;
    }

    const dotThreshold = 0.9995;
    if (d > dotThreshold) {
      // Very close: interpolate linearly and normalise, which avoids dividing by
      // a vanishing sin(angle).
      return Quat(
        x + t * (target.x - x),
        y + t * (target.y - y),
        z + t * (target.z - z),
        w + t * (target.w - w),
      ).normalize();
    }

    final theta0 = math.acos(d);
    final theta = theta0 * t;
    final sinTheta = math.sin(theta);
    final sinTheta0 = math.sin(theta0);

    // `cos(theta)`, which reduces to the standard `sin((1 - t) * theta0) /
    // sin(theta0)`.
    //
    // **This diverges from the Rust source**, which writes `(theta_0 -
    // theta).cos()` with a comment asserting it equals `theta.cos()`. The two
    // agree only at `t == 0.5`, where `theta0 - theta == theta`, and `t == 0.5` is
    // the only value the Rust test covers. At `t == 1.0` the Rust version returns a
    // quaternion two thirds of the way to the target instead of the target, so
    // slerp there does not reach its own endpoint. The Rust `Quat::slerp` needs the
    // same one-token fix.
    final s0 = math.cos(theta) - d * sinTheta / sinTheta0;
    final s1 = sinTheta / sinTheta0;

    return Quat(
      s0 * x + s1 * target.x,
      s0 * y + s1 * target.y,
      s0 * z + s1 * target.z,
      s0 * w + s1 * target.w,
    ).normalize();
  }

  /// Hamilton product: composes two rotations.
  Quat multiply(Quat rhs) => Quat(
        w * rhs.x + x * rhs.w + y * rhs.z - z * rhs.y,
        w * rhs.y - x * rhs.z + y * rhs.w + z * rhs.x,
        w * rhs.z + x * rhs.y - y * rhs.x + z * rhs.w,
        w * rhs.w - x * rhs.x - y * rhs.y - z * rhs.z,
      );

  /// Deliberately minimal: [angularVelocityAsDeltaQuatPerSec] is taken as a
  /// per-second delta rotation and slerped toward by [dtSecs]. Real angular
  /// velocity is an axis-angle vector, which this minimal quaternion cannot
  /// express; for real rotational extrapolation, write the rule for your engine's
  /// quaternion type instead.
  Quat extrapolate(Quat angularVelocityAsDeltaQuatPerSec, double dtSecs) {
    if (angularVelocityAsDeltaQuatPerSec.w.abs() < 1.0 - doubleEpsilon) {
      final scaled = Quat.identity.slerp(angularVelocityAsDeltaQuatPerSec, dtSecs);
      return multiply(scaled).normalize();
    }
    // An identity-like velocity rotates nothing.
    return this;
  }

  @override
  bool operator ==(Object other) =>
      other is Quat && other.x == x && other.y == y && other.z == z && other.w == w;

  @override
  int get hashCode => Object.hash(x, y, z, w);

  @override
  String toString() => 'Quat($x, $y, $z, $w)';
}

/// Straight-line blend between two scalars, the rule to hand a `SnapshotBuffer`
/// of `double`.
double lerpDouble(double a, double b, double t) => a + (b - a) * t;
